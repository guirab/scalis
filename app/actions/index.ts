"use server";
export async function create(formData: NewAccountType) {
  return (
    await fetch("http://localhost:3000/api/accounts", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }))?.json();
}

export async function getAll() {
  return (
    await fetch("http://localhost:3000/api/accounts")
  )?.json();
}

export async function login(formData: LoginType) {
  return await fetch("http://localhost:3000/api/login", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }).then((res:any) => {return res.json()});
}

export async function transferToSameAcc({action, type, amount, id, from}: UpdateType) {
  return await fetch(`http://localhost:3000/api/account/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({action, type, amount, from}),
  }).then((res:any) => {return res.json()});
}

export async function transferToOtherAcc({amount, to, password, id}: TransferOtherAccType) {
  return await fetch(`http://localhost:3000/api/account/transfer/${id}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ amount, to, password }),
  }).then((res:any) => {return res.json()});
}