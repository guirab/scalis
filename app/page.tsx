import Accounts from "./components/accounts";
import NewAccForm from "./components/newAccForm";

export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center p-12 sm:p-24 justify-center sm:justify-start">
      <h1 className="text-5xl">Banking App</h1>
      <div className="w-full flex justify-between flex-col sm:flex-row gap-y-4">
        <NewAccForm />
        <Accounts />
      </div>
    </main>
  );
}
