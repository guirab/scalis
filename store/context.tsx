"use client";

import React, { ReactNode, createContext, useState } from "react";

export const AccountsContext: React.Context<ContextType> = createContext(
  {} as ContextType
);

export const AccountsProvider = ({ children }: { children: ReactNode }) => {
  const [account, setAccount] = useState<AccountType>({} as AccountType);
  const [accounts, setAccounts] = useState<AccountType[]>([] as AccountType[]);

  return (
    <AccountsContext.Provider
      value={{
        account,
        setAccount,
        accounts,
        setAccounts,
      }}
    >
      {children}
    </AccountsContext.Provider>
  );
};
