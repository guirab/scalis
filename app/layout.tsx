import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

import { AccountsProvider } from "@/store/context";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Banking App",
  description: "Banking App for SCALIS",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <AccountsProvider>{children}</AccountsProvider>
      </body>
    </html>
  );
}
