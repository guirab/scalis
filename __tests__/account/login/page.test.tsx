import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { AccountsContext } from "../../../store/context";
import Login from "../../../app/account/login/page";
import { login } from "../../../app/actions";
import { useRouter } from "next/navigation";
import { afterEach, beforeAll, describe, expect, test, vitest } from "vitest";

global.fetch = vitest.fn();

function createFetchResponse(data: any) {
  return { json: () => new Promise((resolve) => resolve(data)) };
}

describe("Login", () => {
  afterEach(() => {
    cleanup();
  });
  beforeAll(() => {
    vitest.mock("next/navigation", () => ({
      useRouter: () => ({
        back: vitest.fn(),
      }),
    }));
    vitest.mock("react", async () => {
      return {
        ...(await vitest.importActual("react")),
        useContext: () => ({
          account: {
            username: "testuser",
            checking: 1000,
            savings: 2000,
          },
          setAccount: vitest.fn(),
        }),
      };
    });
  });

  test("snapshot", () => {
    const { container } = render(<Login />);
    expect(container).toMatchSnapshot();
  });

  test("renders login form correctly", () => {
    render(<Login />);
    expect(screen.getAllByText("Login")[0]).not.toBeNull();
    expect(screen.getByLabelText("Username")).not.toBeNull();
    expect(screen.getByLabelText("Password")).not.toBeNull();
    expect(screen.getByRole("button", { name: "Login" })).not.toBeNull();
  });

  test("calls login function and redirects to account page on successful login", async () => {
    render(<Login />);
    const payload = {
      username: "testuser",
      password: "password",
    };
    const usernameInput = screen.getByLabelText("Username");
    const passwordInput = screen.getByLabelText("Password");
    const loginButton = screen.getByRole("button", { name: "Login" });

    fireEvent.change(usernameInput, { target: { value: "testuser" } });
    fireEvent.change(passwordInput, { target: { value: "password" } });
    fireEvent.click(loginButton);

    (fetch as any).mockResolvedValue(createFetchResponse({}));

    const loginData = await login(payload as any);

    expect(fetch).toHaveBeenCalledWith("http://localhost:3000/api/login", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        username: "testuser",
        password: "password",
      }),
    });
    expect(loginData).toStrictEqual({});
  });
});
