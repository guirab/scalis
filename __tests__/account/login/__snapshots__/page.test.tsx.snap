// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > snapshot 1`] = `
<div>
  <main
    class="flex min-h-screen flex-col items-center relative p-12 sm:p-24 justify-center sm:justify-start"
  >
    <div
      class="w-full pb-12"
    >
      <span
        class="flex items-center text-2xl cursor-pointer w-fit"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M427 234.625H167.296l119.702-119.702L256 85 85 256l171 171 29.922-29.924-118.626-119.701H427v-42.75z"
          />
        </svg>
        Back
      </span>
    </div>
    <div
      class="bg-white bg-opacity-30 rounded-lg h-fit w-full p-4"
    >
      <h1
        class="text-5xl"
      >
        Login
      </h1>
      <form
        class="flex flex-col w-full gap-y-4 mt-4"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="off"
          class="text-black pl-2"
          id="username"
          name="username"
          required=""
          type="text"
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="off"
          class="text-black pl-2"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <button
          class="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </main>
</div>
`;
